### PR TITLE
Add DeepSparse Evaluator

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds a new Evaluator class in DeepSparse to evaluate
onnx models.

Right now we only support perplexity calculation with LLMS
Future support will include running with custom evaluation functions

Usage:
```python
from deepsparse.transformers.evaluator import TransformersEvaluator


model = "/home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment"
dataset = "openai_humaneval"
evaluator = TransformersEvaluator(model=model, dataset=dataset)
evaluator.evaluate()
print(evaluator.get_results())
```

Output:
```bash
/home/rahul/projects/.venv/bin/python /home/rahul/projects/deepsparse/local-run.py                                            (evaluator-api|●1✚3…8⚑3)
2023-09-14 09:53:27 deepsparse.transformers WARNING  The neuralmagic fork of transformers may not be installed. It can be installed via `pip install nm_transformers`
Found cached dataset openai_humaneval (/home/rahul/.cache/huggingface/datasets/openai_humaneval/openai_humaneval/1.0.0/2955cebd73602e828fa8c0a424c594e5fab4ec863b316ca98f3d8fdb6a626e75)
Using pad_token, but it is not set yet.
2023-09-14 09:53:30 deepsparse.transformers.pipelines.text_generation INFO     Compiling an auxiliary engine to process a prompt with a larger processing length. This improves performance, but may result in additional memory consumption.
2023-09-14 09:53:32 deepsparse.utils.onnx INFO     Overwriting in-place the input shapes of the transformer model at /home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment/model.onnx
DeepSparse, Copyright 2021-present / Neuralmagic, Inc. version: 1.6.0.20230910 COMMUNITY | (c66b57da) (release) (optimized) (system=avx512, binary=avx512)
2023-09-14 09:53:40 deepsparse.utils.onnx INFO     Overwriting in-place the input shapes of the transformer model at /home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment/model.onnx
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 164/164 [04:00<00:00,  1.47s/it]
{'perplexities': array([ 2.4683068,  3.4789572,  7.24103  ,  4.920437 ,  3.6537664,
        2.6923187,  3.8851712,  2.9729612,  2.4295907,  2.9990573,
        3.1750627,  3.01282  ,  2.6954668,  2.0642262,  2.6666799,
        2.8861017,  3.786744 ,  4.641316 ,  2.6624188,  2.4186363,
        2.0480666,  2.5479674,  3.402345 ,  2.7455387,  3.5062313,
        3.077313 ,  2.8929992,  3.6263716,  2.0716732,  2.7997034,
        3.215443 ,  2.122958 ,  3.6990309,  4.0144134,  3.785419 ,
        2.917665 ,  2.9874372,  3.0300252,  3.9524636,  2.0319357,
        2.049934 , 13.171704 ,  3.2868052,  2.260889 ,  3.459201 ,
        5.8242908,  2.8138506,  2.3460915,  1.9261636,  3.628064 ,
        2.5298882,  2.8170197,  2.9886558,  2.5365953,  2.8191276,
        1.965102 ,  2.8490472,  2.792732 ,  3.407229 ,  2.3122377,
        2.5336177,  2.5025997,  4.160729 ,  2.5276237,  3.26036  ,
        3.1124425,  6.955526 ,  4.9945974,  3.97927  ,  3.093798 ,
        3.2788186,  2.866415 ,  3.0416899,  2.9682531,  2.5747738,
        2.961615 ,  2.8990846,  4.1413093,  4.1894035,  5.053281 ,
        2.5333493,  4.285258 ,  2.8351367,  6.271922 ,  6.207515 ,
        7.3686795,  6.165307 ,  3.6138704,  3.9438305,  4.36005  ,
        2.67872  ,  5.695831 ,  3.0526283,  5.1900773,  5.0284314,
        2.566863 ,  2.0414267,  6.171901 ,  3.181557 ,  3.433489 ,
        4.56728  ,  2.2250726,  3.5549545,  3.4682872,  4.3366594,
        3.0210724,  3.0803595,  2.3528535,  3.754858 ,  4.8585315,
        2.9731972,  3.0983877,  4.438951 ,  5.238712 ,  3.121741 ,
        3.6424196,  3.4360666,  3.4695528,  3.4752402,  3.3812926,
        2.5247307,  5.615052 ,  5.7387547,  3.9792697,  2.9349108,
        4.247634 ,  2.111962 ,  3.8231397,  4.674957 ,  5.8530736,
        4.0250497,  3.3499212,  2.7815971,  3.8335645,  3.0472577,
        3.4348195,  2.7136924,  3.521654 ,  3.605058 ,  3.4651494,
        2.3944056,  3.5791583,  3.5510464,  3.0122628,  3.781657 ,
        3.8275018,  3.5090864,  2.6653678,  2.9007623,  4.4009748,
        3.6951585,  3.689766 ,  7.8209844,  6.013169 ,  4.826395 ,
        2.6200743,  3.006651 ,  3.3689177,  5.2386904,  5.287668 ,
        4.7331705,  4.1493807,  5.70622  ,  2.9901416], dtype=float32), 'mean_perplexity': 3.635836}
```
 
Note: Support to run evaluation across nodes will be added as a part of #1243 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205386356840761

### CODE REVIEW

File: src/deepsparse/transformers/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    @abstractmethod
+    def get_results(self):
+        """
+        Return the results of the evaluation.
+        """
+        raise NotImplementedError
+
+
+class LMEvaluator(Evaluator):
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/evaluator.py:60]: LM stands for Language Models? So we can evalute both "base" transformers as well as LLMs?

**Review by dbogunowicz — APPROVED** [review]

File: src/deepsparse/transformers/evaluator.py (hunk 1 above)
Comment (rahul-tuli) [reply src/deepsparse/transformers/evaluator.py:60]: Yes Perplexity can be calculated for all LMs

Comment (dbogunowicz) [reply src/deepsparse/transformers/evaluator.py:60]: I am just not sure about the term LM. May not be obvious what this means. Can we just rename it to TransformersEvaluator or NLPEvaluator?

Comment (rahul-tuli) [reply src/deepsparse/transformers/evaluator.py:60]: Do you think PerplexityEvaluator is better? as this does not work with all Transformers/all NLP models?

Comment (dbogunowicz) [reply src/deepsparse/transformers/evaluator.py:60]: Perplexity works for all the Transformer models as far as I know. But the cool thing about your evaluator is that you can extend it to other "LLM-type" metrics like pass@k.

Comment (dbogunowicz) [reply src/deepsparse/transformers/evaluator.py:60]: `TransformersEvaluator` sounds like the best choice for me imo.

Comment (rahul-tuli) [reply src/deepsparse/transformers/evaluator.py:60]: Accepted!

**Review by bfineran — DISMISSED** [review]
LGTM - let's add in these simple UX changes per request from product

```
from deepsparse import evaluate, Evaluator

result = evaluate(pipeline=PIPELINE, data=DATA, metrics=METRICS, **kwargs)
result = Evaluator(pipeline=PIPELINE, **kwargs)(data=DATA, metrics=METRICS, **kwargs)
```
```
deepsparse.eval --model MODEL --data DATA --metrics METRICS **kwargs
```

**Comment by dbogunowicz** [comment]
@bfineran @rahul-tuli we can probably close this.

**Comment by rahul-tuli** [comment]
Closed in favor of new Evaluation Harness #1382

**Pull request closed without merging**